Implementing MetaData upload API

saurabh1969 opened the pull request:
MetaData Upload API endpoints implementation as per #4 

Implemented only for experiments,sites and citations.

saurabh1969 committed:
Update yaba.yaml

saurabh1969 committed:
Implementing Endpoints for YABA API

dlebauer commented:
@saurabh1969 could you please flush out the README and include the instructions for deploying the API locally? This might be easier to do with Docker. Should that be a separate issue?

In the README, could you please provide examples for inserting data, and links to sample csv and shape files?

saurabh1969 committed:
Updated credentials

saurabh1969 committed:
Updated exception handling

saurabh1969 committed:
Updated sites endpoint

dlebauer committed:
Update yaba.yaml

dlebauer committed:
Update yaba.yaml

saurabh1969 committed:
Updated experiments,sites,cultivars enpoints

saurabh1969 committed:
Updated README file and docker-compose

saurabh1969 committed:
Updated README

saurabh1969 committed:
Updated Dockerfile

saurabh1969 committed:
Updated dockerfile

saurabh1969 committed:
Added experiments_sites and experiments_treatments endpoints.

saurabh1969 committed:
Updated with new endpoints

saurabh1969 committed:
Updated with treatments and citations endpoints

saurabh1969 committed:
Added Sites_cultivars and Citations_sites endpoints.

saurabh1969 committed:
Updated Docker-compose file.

Before setup for bety was tedious.This docker-compose file reduces the steps and running the app becomes easy.

saurabh1969 committed:
Updated docker-compose file

saurabh1969 committed:
Exception handling added.

Exception handling for Integrity or Constraint error,Operational error,Wrong file input and other exceptions.
Now user will receive appropriate message when such conditions will arise.

dlebauer commented:
@kimberlyh66 could you make sure to walk through the readme and confirm that these work with a sample dataset?

dlebauer commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+## How to hit the endpoints
+
+Following endpoints can be used to upload Metadata to respective tables in bety.
```
How do we know that the insert is successful? What is the expected output?

dlebauer commented on README.md:
````diff
@@ -1,2 +1,105 @@
+
+```sh
+curl -F "fileName=@input_files/experiments.csv"   \
+     http://localhost:5001/yaba/v1/experiments?username=guestuser
````
This is returning code 402 and no other information.... should say something like "successfully inserted $n records to $table table"

kimberlyh66 replied:
Returning a 402 - not sure what the error is

saurabh1969 committed:
Updated IP address

kimberlyh66 commented on README.md:
````diff
@@ -1,2 +1,105 @@
+Sites: (to sites table)
+
+```sh
+curl -F "fileName=@sites.csv"   \
````
missing `input_files` part of path to sites.csv,
getting a 410 when running

kimberlyh66 commented on README.md:
````diff
@@ -1,2 +1,105 @@
+Treatments: (to treatments table)
+
+```sh
+curl -F "fileName=@input_files/treatments.csv"   \
````
Returns a 402

kimberlyh66 commented on README.md:
````diff
@@ -1,2 +1,105 @@
+Cultivars: (to cultivars table)
+
+```sh
+curl -F "fileName=@input_files/cultivars.csv"   \
````
Returns a 402

kimberlyh66 commented on README.md:
````diff
@@ -1,2 +1,105 @@
+Citations: (to citations table)
+
+```sh
+curl -F "fileName=@input_files/citations.csv"   \
````
Returns a 402

kimberlyh66 reviewed (changes requested):
I am getting the code 402 when trying to upload data to the bety tables. I am not sure if I am connected to the database correctly?

saurabh1969 committed:
With tests and Travis yml file.

saurabh1969 committed:
Update .travis yml file

saurabh1969 committed:
Updated the file location in travis.yml

saurabh1969 committed:
Updated

saurabh1969 committed:
Update requirements.txt

saurabh1969 committed:
Updated app

saurabh1969 committed:
Updated

KristinaRiemer commented:
@kimberlyh66 your 402 errors have to doing with a PGHOST setting need to be set manually to your IP address, which @saurabh1969 is currently trying to fix.

saurabh1969 commented:
@dlebauer @kimberlyh66 your 402 error is valid point.It is do with PGHOST in credentials.In Docker,the app is getting connected to bety.So it requires the host network,which is different for every user.So i am just trying to figuring out that issue.For the time being, **docker network inspect bridge | grep "Gateway"**.You will get something like this - **"Gateway": "172.17.0.1"**. Your can put the in PGHOST variable in credentials.yaml inside app folder.Then do docker-compose up.

saurabh1969 committed:
Updated app with new error codes and sites endpoint

saurabh1969 committed:
Docker PGHOST issue fixed and response for each endpoint enhanced

saurabh1969 committed:
Signed-off-by: saurabh1969 <kumarsaurabh772@gmail.com>

saurabh1969 committed:
Update README.md

Putting badge Icon

saurabh1969 commented:
@kimberlyh66 @dlebauer @KristinaRiemer PGHOST issue fixed.Now you can clone the repo and can do docker-compose up.

Chris-Schnaufer commented on app/Meta.py:
```diff
@@ -0,0 +1,564 @@
+    a_file_name = secure_filename(File.filename)
+
+    # Build target
+    a_file_target = os.path.join(os.getcwd(),UPLOAD_FOLDER,a_file_name)
```
Suggest moving the construct 'os.getcwd(),UPLOAD_FOLDER' to where UPLOAD_FOLDER is defined; perhaps something like UPLOAD_PATH=os.path.join(os.getcwd(),UPLOAD_FOLDER)

Chris-Schnaufer commented on app/Meta.py:
```diff
@@ -0,0 +1,564 @@
+                    410 Default error.See logs for more information
+    """
+    try:
+        save_tempFile(shp_file)
```
I don't believe the temporary files in the new location will auto-delete. They should be deleted before the function returns - perhaps in a finally block

Chris-Schnaufer commented on app/Meta.py:
```diff
@@ -0,0 +1,564 @@
+            data.iat[i,6]=p.wkb_hex
+
+        #Checking necessary columns are there.
+        columns=data.columns.values.tolist()
```
I suggest moving this check to immediately after the csv file is loaded (removing the 'geometry' column that's added later) and immediately returning if the check fails.

Chris-Schnaufer commented on app/Meta.py:
```diff
@@ -0,0 +1,564 @@
+    try:
+        user_id=fetch_id(username,table='users')
+
+        if user_id =='':
```
For this check, `if not user_id:` might be preferred since it'll work with None, empty strings, etc. Maintainers don't need to worry about the fetch_id return type

Chris-Schnaufer commented on app/Meta.py:
```diff
@@ -0,0 +1,564 @@
+        if user_id =='':
+            return 401
+        #Checking necessary columns are there.
+        columns=data.columns.values.tolist()
```
data is not loaded yet

Chris-Schnaufer commented on app/db.py:
```diff
@@ -0,0 +1,86 @@
+		connection.commit()
+     
+
+def fetch_specie_id(species):
```
Consider using lambda language feature to reduce the repetition for fetching IDs

saurabh1969 replied:
Implemention done.

saurabh1969 committed:
Updated site endpoint and .travis yml

saurabh1969 committed:
Update requirements.txt

saurabh1969 committed:
Update .travis.yml

saurabh1969 committed:
Update .travis.yml

saurabh1969 committed:
Updated app as per Chris recommendations

saurabh1969 committed:
Implemented Lambda fuctions where required and introduced the Client Library

saurabh1969 committed:
Implemented Client Library

saurabh1969 committed:
Implemented lambda functions

saurabh1969 committed:
Sites endpoint implemented and test file.

saurabh1969 committed:
Updated requirement file (removing unittest)

saurabh1969 committed:
Updated test BASE url

saurabh1969 committed:
Updated insert_sites test method

saurabh1969 committed:
Travis recheck-1

saurabh1969 committed:
Travis check-2

saurabh1969 committed:
Travis test-3

saurabh1969 committed:
Travis check 4

saurabh1969 committed:
Travis check 5

saurabh1969 committed:
Travis check 6

saurabh1969 committed:
Travis check 7

dlebauer commented:
Please open two new PRs so that we can conduct separate reviews of the API and the client

Once those PRs are open, you can close this one.
